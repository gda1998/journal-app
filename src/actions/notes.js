import { db } from '../firebase/firebaseConfig';
import types from '../types/types';
import {
    addDoc,
    collection,
    doc,
    updateDoc,
    deleteDoc,
    // getDoc
} from 'firebase/firestore';
import Swal from 'sweetalert2';
import { swalAlert, swalLoading } from '../helpers/swalToast';
import loadNotes from '../helpers/loadNotes';
import fileUpload from '../helpers/fileUpload';
// import fileDelete from '../helpers/fileDelete';

export const startNewNoteAction = () => {
    return (dispatch, getState) => {
        const { uid } = getState().auth;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        };

        const notesRef = collection(db, `${uid}/journal/notes`);
        return addDoc(notesRef, newNote)
            .then(docRef => {
                dispatch(activeNoteAction(docRef.id, newNote));
                dispatch(addNewNoteAction(docRef.id, newNote));
            })
            .catch(console.error);
    }
}

export const activeNoteAction = (id, note) => ({
    type: types.notesActive,
    payload: { id, ...note }
});

export const addNewNoteAction = (id, note) => ({
    type: types.notesAddNew,
    payload: { id, ...note }
});

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotesAction(notes));
    }
}

export const setNotesAction = (notes) => ({
    type: types.notesLoad,
    payload: notes
});

export const startUpdateNote = (note) => {
    return (dispatch, getState) => {
        const { uid } = getState().auth;
        !note.url && (note.url = null);
        const noteRef = doc(db, `${uid}/journal/notes/`, note.id);
        updateDoc(noteRef, note);
        dispatch(refreshNoteAction(note));
        swalAlert('Updated!', 'The note was updated', 'success');
    }
}

export const refreshNoteAction = (note) => ({
    type: types.notesUpdate,
    payload: note
});

export const startUploadingImgAction = (file) => {
    return async (dispatch, getState) => {
        const { active: activeNote } = getState().notes;
        swalLoading('Uploading', 'Please wait...');

        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;

        Swal.close();
        await dispatch(startUpdateNote(activeNote));
    }
}

export const startDeleteNote = (id) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        try {
            const ref = await doc(db, uid, `journal/notes/${id}`);
            // const noteRef = await getDoc(ref);

            // Delete the image from cloudinary and from the db
            // fileDelete(noteRef.data().url)
            await deleteDoc(ref);

            dispatch(deleteNote(id));
        } catch (error) {
            console.error('Error al eliminar la nota: ', error);
        }
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
});

export const notesLogout = () => ({
    type: types.notesLogoutCleaning
})