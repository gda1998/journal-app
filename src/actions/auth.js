import { 
    createUserWithEmailAndPassword,
    getAuth,
    signInWithPopup,
    signInWithEmailAndPassword,
    signOut,
    updateProfile 
} from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebaseConfig';
import { finishLoadingAction, startLoadingAction } from './ui';
import { swalAlert } from '../helpers/swalToast';
import types from '../types/types';
import { notesLogout } from './notes';

export const startRegisterAction = ({ email, password, name }) => {
    return (dispatch) => {
        const auth = getAuth();
        dispatch( startLoadingAction() );
        createUserWithEmailAndPassword(auth, email, password)
        .then( async({ user }) => {
            await updateProfile(user, { displayName: name });
            dispatch( login(user.uid, user.displayName) );
        })
        .catch( e => swalAlert('Error!', e, 'error') )
        .finally( () => dispatch( finishLoadingAction() ) );
    };
};

export const startLoginEmailPassword = ({ email, password }) => {
    return (dispatch) => {
        const auth = getAuth();
        dispatch( startLoadingAction() );
        return signInWithEmailAndPassword(auth, email, password)
        .then( ({user}) => dispatch( login(user.uid, user.displayName) ) )
        .catch( e => swalAlert('Error!', e, 'error') )
        .finally( () => dispatch( finishLoadingAction() ) );
    };
};

export const startGoogleLogin = () => {
    return (dispatch) => {
        const auth = getAuth();
        signInWithPopup(auth, googleAuthProvider)
        .then( ({user}) => {
            dispatch( login(user.uid, user.displayName) )
        });
    }
}

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

export const logout = () => ({
    type: types.logout
})

export const startLogout = () => {
    return async(dispatch) => {
        await signOut( getAuth() );
        dispatch( logout() );
        dispatch(  notesLogout() );
    }
}