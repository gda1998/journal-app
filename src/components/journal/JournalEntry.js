import dayjs from 'dayjs';
import { useDispatch } from 'react-redux';
import { activeNoteAction } from '../../actions/notes';


export const JournalEntry = ({ id, title, body, date, url }) => {
    const noteDate = dayjs(date);
    const dispatch = useDispatch();
    const handleEntryClick = () => dispatch( activeNoteAction(id, { title, body, date, url }) );

    return (
        <div 
            className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
            onClick={ handleEntryClick }
        >

            { url && 
                <div 
                    className="journal__entry-picture"
                    style={{ 
                        backgroundSize: 'cover',
                        backgroundImage: `url(${ url })`
                    }}
                >
                </div>/* /.journal__entry-picture */
            }


            <div className="journal__entry-body">
                <p className="journal__entry-title">
                    { title }
                </p>
                <p className="journal__entry-content">
                    { body }
                </p>
            </div>{/* /.journal__entry-body */}

            <div className="journal__entry-date-box">
                <span>{ noteDate.format('dddd') }</span>
                <h4>{ noteDate.format('DD') }</h4>
            </div>{/* /.journal__entry-date-box */}

        </div>/* /.journal__entry */
    );
}