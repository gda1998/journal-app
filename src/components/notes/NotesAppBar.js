import { useDispatch, useSelector } from 'react-redux';
import { startUpdateNote, startUploadingImgAction } from '../../actions/notes';

export const NotesAppBar = () => {
    // Redux
    const dispatch = useDispatch();
    const { active:note } = useSelector( state => state.notes );
    const handleUpdate = () => dispatch( startUpdateNote(note) );
    const handlePictureClick = () => document.querySelector('#fileSelector').click();
    
    const handleFileChange = ({target}) => {
        const file = target.files[0];
        file && dispatch( startUploadingImgAction(file) );
    }

    return (
        <div className="notes__appbar">
            <span>05 de Noviembre de 2021</span>

            <input
                id="fileSelector"
                onChange ={ handleFileChange }
                type="file"
                style={{ display: 'none' }}
            />

            <div>
                <button 
                    className="btn"
                    onClick={ handlePictureClick }
                >
                    Picture
                </button>
                <button 
                    className="btn"
                    onClick={ handleUpdate }
                >
                    Save
                </button>
            </div>
        </div>/* /.notes__appbar */
    );
}