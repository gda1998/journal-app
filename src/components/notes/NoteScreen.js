import { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar';
import { activeNoteAction, startDeleteNote } from '../../actions/notes';

export const NoteScreen = () => {
    // Redux
    const dispatch = useDispatch();
    const { active:note } = useSelector( state => state.notes );

    // useForm
    const [ formValues, handleInputChange, reset ] = useForm(note);
    const { title, body } = formValues;
    note.url && ( formValues.url = note.url );

    // useRef
    const activeId = useRef(note.id);

    // handleDelete
    const handleDelete = () => dispatch( startDeleteNote(note.id) );

    useEffect(() => {
        if( activeId.current !== note.id ){
            reset(note);
            activeId.current = note.id;
        }
    }, [note, reset]);

    useEffect(() => {
        dispatch( activeNoteAction(formValues.id, {...formValues}) );
    }, [dispatch, formValues]);
    

    return (
        <div className="notes__main-content">

            <NotesAppBar />
            <div className="notes__content">

                <input
                    autoComplete="off"
                    className="notes__title-input"
                    name="title"
                    onChange={ handleInputChange }
                    placeholder="Some awesome title"
                    type="text"
                    value={ title }
                />
                <textarea
                    className="notes__textarea"
                    name="body"
                    onChange={ handleInputChange }
                    placeholder="What's happened today?"
                    type="text"
                    value={ body }
                ></textarea>

                {
                    note.url &&
                    <div className="notes__image">
                        <img
                            src={ note.url }
                            alt="imagen"
                        />
                    </div>/* /.notes__image */
                }

            </div>{/* /.notes__content */}

            <button
                className="btn btn-danger"
                onClick={ handleDelete }
            >
                Delete
            </button>

        </div>/* /.notes__main-content */
    );
}
