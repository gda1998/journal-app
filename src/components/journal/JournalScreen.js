import { useSelector } from 'react-redux';
import { Navigate } from 'react-router-dom';
import { NoteScreen } from '../notes/NoteScreen';
import { Sidebar } from './Sidebar';
import { NothingSelected } from './NothingSelected';

export const JournalScreen = () => {
    const { auth, notes } = useSelector( state => state );
    return Object.keys(auth).length === 0 
    ? <Navigate to="/auth/login" />
    : (
        <div className="journal__main-content animate__animated animate__fadeIn animate__faster">
            <Sidebar />
            <main>
                {
                    notes.active === null ? <NothingSelected /> : <NoteScreen />
                }
            </main>
        </div>
    );
}
