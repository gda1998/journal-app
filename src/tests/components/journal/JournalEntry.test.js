import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { JournalEntry } from '../../../components/journal/JournalEntry';
import { activeNoteAction } from '../../../actions/notes';

const mockStore = configureStore([thunk]);
const store = mockStore({});
store.dispatch = jest.fn();

const note = {
    id: '1234',
    title: 'I am a title',
    body: 'I am the body',
    date: '13859454',
    url: 'https://any.com'
}

describe('Prueba en <JournalEntry />', () => {
    let wrapper;
    const getWrapper = (store) => mount(
        <Provider store={store}>
            <JournalEntry {...note} />
        </Provider>
    );

    beforeEach( () => wrapper = getWrapper(store));

    test('Debe de mostrarse correctamente', () => { 
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de llamarse el action activeNoteAction', () => { 
        wrapper.find('.journal__entry').simulate('click');
        expect(store.dispatch).toHaveBeenCalledWith(
            activeNoteAction(note.id, note)
        );
    });
});