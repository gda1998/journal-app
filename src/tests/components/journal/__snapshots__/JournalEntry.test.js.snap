// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en <JournalEntry /> Debe de mostrarse correctamente 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://any.com)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      I am a title
    </p>
    <p
      className="journal__entry-content"
    >
      I am the body
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Friday
    </span>
    <h4>
      23
    </h4>
  </div>
</div>
`;
